rmf_site_editor/src/widgets/canvas_tooltips.rs

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+use bevy::ecs::prelude::Resource;
+use bevy_egui::egui::{self, Context};
+use std::borrow::Cow;
+
+/// This resource holds a list of tooltips that are emitted by 3D canvas interactions
+/// and should be rendered near the cursor. Each tip will be given its own line in a
+/// text block.
+///
+/// The `tips` field is cleared out on each iteration to prevent sticky tooltips.
+#[derive(Resource, Default)]
+pub struct CanvasTooltips {
+    /// Push a `Cow::Borrowed("Your string literal here")` to push a `&'static str`,
+    /// which is the most efficient way to produce tooltip text.
+    ///
+    /// If your text is not a string literal then push `Cow::Owned(your_string)`.
+    tips: Vec<Cow<'static, str>>,
+    /// We keep track of previous tooltips to keep order as consistent as possible
+    /// to prevent flickering as it's possible for different systems to add in
+    /// tooltips at different times.
+    previous: Vec<Cow<'static, str>>,
+}
```

Comment:
A bit nitpicky I guess but what if the tooltips were just in an automatically ordered structure (i.e. `BTreeSet`)? It sounds like ther order of pushing is a bit erratic anyway so maybe a deterministic alphabetical order would just get rid of the complexity of keeping track of the old values and doing searches every iteration

Reply:
The rationale for the current approach is that the oldest items in the tooltip will remain at the top, which I think gives better stability for a human who may be trying to read it. If we use a collection with built-in ordering like `BTreeSet` then new tips can, at any time, shift down a message that a user is paying attention to. Messages that are more fickle will remain towards the bottom, which I think is also helpful for the user to be able to focus on the bottom if they want to watch what tooltips are coming and going.

Personally I think this (admittedly subtle) quality of life improvement is worth the added complexity, but what do you think now that I've explained the rationale?

Reply:
I see, I thought the main problem to be solved was that the tooltips were usually the same but system ordering might cause them to be pushed in different order, but if there is frequent adding / removing this sounds good